File: src/jsMain/kotlin/kingdom/lite/model/PlayerActions.kt
package kingdom.lite.model

import kotlin.js.Json

/**
 * Data model for Player Actions loaded from JSON files
 */

data class SkillOption(
    val skill: String,
    val description: String
)

data class ActionEffect(
    val description: String,
    val modifiers: Map<String, Any> = emptyMap()
)

data class PlayerAction(
    val id: String,
    val name: String,
    val category: String,
    val description: String,
    val skills: List<SkillOption> = emptyList(),
    val criticalSuccess: ActionEffect,
    val success: ActionEffect,
    val failure: ActionEffect,
    val criticalFailure: ActionEffect,
    val proficiencyScaling: Map<String, Int>? = null,
    val special: String? = null,
    val cost: Map<String, Int>? = null  // For actions that have resource costs
)

object PlayerActionsData {
    // Category display names
    val categoryNames = mapOf(
        "uphold-stability" to "Uphold Stability",
        "military-operations" to "Military Operations", 
        "expand-borders" to "Expand the Borders",
        "urban-planning" to "Urban Planning",
        "foreign-affairs" to "Foreign Affairs",
        "economic-actions" to "Economic Actions"
    )
    
    val categoryDescriptions = mapOf(
        "uphold-stability" to "Maintain the kingdom's cohesion by resolving crises and quelling unrest.",
        "military-operations" to "War must be waged with steel and strategy.",
        "expand-borders" to "Seize new territory to grow your influence and resources.",
        "urban-planning" to "Your people need places to live, work, trade, and worship.",
        "foreign-affairs" to "No kingdom stands alone.",
        "economic-actions" to "Manage trade and personal wealth."
    )
    
    // Mock data for now - in production this would load from JSON files
    fun getAllActions(): List<PlayerAction> {
        // This will be replaced with actual JSON loading
        return listOf(
            // Expand Borders actions
            PlayerAction(
                id = "claim-hexes",
                name = "Claim Hexes",
                category = "expand-borders",
                description = "Assert sovereignty over new territories, expanding your kingdom's borders into unclaimed lands",
                skills = listOf(
                    SkillOption("nature", "harmonize with the land"),
                    SkillOption("survival", "establish frontier camps"),
                    SkillOption("intimidation", "force submission"),
                    SkillOption("occultism", "mystical claiming rituals"),
                    SkillOption("religion", "divine mandate")
                ),
                criticalSuccess = ActionEffect("Claim all targeted hexes +1 extra hex", mapOf("hexesClaimed" to "proficiency+1")),
                success = ActionEffect("Claim targeted hexes (based on proficiency)", mapOf("hexesClaimed" to "proficiency")),
                failure = ActionEffect("No effect"),
                criticalFailure = ActionEffect("No effect"),
                proficiencyScaling = mapOf("trained" to 1, "expert" to 1, "master" to 2, "legendary" to 3),
                special = "+2 circumstance bonus when claiming hexes adjacent to 3+ controlled hexes"
            ),
            PlayerAction(
                id = "build-roads",
                name = "Build Roads",
                category = "expand-borders",
                description = "Construct pathways between settlements to improve trade, travel, and military movement",
                skills = listOf(
                    SkillOption("crafting", "engineering expertise"),
                    SkillOption("survival", "pathfinding routes"),
                    SkillOption("athletics", "manual labor"),
                    SkillOption("nature", "work with terrain")
                ),
                criticalSuccess = ActionEffect("Build roads +1 hex", mapOf("roadsBuilt" to 2)),
                success = ActionEffect("Build roads", mapOf("roadsBuilt" to 1)),
                failure = ActionEffect("No effect"),
                criticalFailure = ActionEffect("No effect")
            ),
            PlayerAction(
                id = "send-scouts",
                name = "Send Scouts",
                category = "expand-borders",
                description = "Learn about unexplored hexes",
                skills = listOf(
                    SkillOption("survival", "wilderness navigation"),
                    SkillOption("nature", "track and explore")
                ),
                criticalSuccess = ActionEffect("Scout 2 hexes"),
                success = ActionEffect("Scout 1 hex"),
                failure = ActionEffect("No effect"),
                criticalFailure = ActionEffect("No effect")
            ),
            PlayerAction(
                id = "fortify-hex",
                name = "Fortify Hex",
                category = "expand-borders",
                description = "Strengthen defensive positions",
                skills = listOf(
                    SkillOption("crafting", "build fortifications"),
                    SkillOption("warfare", "strategic placement")
                ),
                criticalSuccess = ActionEffect("Fortify hex with bonus"),
                success = ActionEffect("Fortify hex"),
                failure = ActionEffect("No effect"),
                criticalFailure = ActionEffect("No effect")
            ),
            PlayerAction(
                id = "create-worksite",
                name = "Create Worksite",
                category = "expand-borders",
                description = "Establish farms, mines, quarries, or lumber camps",
                skills = listOf(
                    SkillOption("crafting", "build infrastructure"),
                    SkillOption("survival", "identify resources")
                ),
                criticalSuccess = ActionEffect("Create worksite with bonus"),
                success = ActionEffect("Create worksite"),
                failure = ActionEffect("No effect"),
                criticalFailure = ActionEffect("No effect")
            ),
            
            // Uphold Stability actions
            PlayerAction(
                id = "coordinated-effort",
                name = "Coordinated Effort",
                category = "uphold-stability",
                description = "Two PCs work together on a single action with a bonus",
                skills = listOf(
                    SkillOption("diplomacy", "coordinate efforts"),
                    SkillOption("society", "organize teamwork")
                ),
                criticalSuccess = ActionEffect("Action succeeds with major bonus"),
                success = ActionEffect("Action succeeds with bonus"),
                failure = ActionEffect("No effect"),
                criticalFailure = ActionEffect("No effect")
            ),
            PlayerAction(
                id = "arrest-dissidents",
                name = "Arrest Dissidents",
                category = "uphold-stability",
                description = "Convert unrest into imprisoned unrest",
                skills = listOf(
                    SkillOption("intimidation", "suppress dissent"),
                    SkillOption("society", "identify troublemakers")
                ),
                criticalSuccess = ActionEffect("Convert 2 unrest to imprisoned"),
                success = ActionEffect("Convert 1 unrest to imprisoned"),
                failure = ActionEffect("No effect"),
                criticalFailure = ActionEffect("Gain 1 unrest")
            ),
            PlayerAction(
                id = "deal-with-unrest",
                name = "Deal with Unrest",
                category = "uphold-stability",
                description = "Directly reduce unrest by 1-3 based on success",
                skills = listOf(
                    SkillOption("diplomacy", "calm the populace"),
                    SkillOption("performance", "inspire the people"),
                    SkillOption("intimidation", "quell dissent")
                ),
                criticalSuccess = ActionEffect("Reduce unrest by 3"),
                success = ActionEffect("Reduce unrest by 1"),
                failure = ActionEffect("No effect"),
                criticalFailure = ActionEffect("Gain 1 unrest")
            ),
            
            // Military Operations
            PlayerAction(
                id = "recruit-unit",
                name = "Recruit a Unit",
                category = "military-operations",
                description = "Raise new troops for your armies",
                skills = listOf(
                    SkillOption("warfare", "military recruitment"),
                    SkillOption("intimidation", "conscription")
                ),
                criticalSuccess = ActionEffect("Recruit unit with bonus"),
                success = ActionEffect("Recruit unit"),
                failure = ActionEffect("No effect"),
                criticalFailure = ActionEffect("No effect")
            ),
            PlayerAction(
                id = "deploy-army",
                name = "Deploy Army",
                category = "military-operations",
                description = "Move troops to strategic positions",
                skills = listOf(
                    SkillOption("warfare", "tactical deployment"),
                    SkillOption("survival", "logistics")
                ),
                criticalSuccess = ActionEffect("Deploy with extra movement"),
                success = ActionEffect("Deploy army"),
                failure = ActionEffect("No effect"),
                criticalFailure = ActionEffect("No effect")
            ),
            
            // Urban Planning
            PlayerAction(
                id = "establish-settlement",
                name = "Establish a Settlement",
                category = "urban-planning",
                description = "Found a new village",
                skills = listOf(
                    SkillOption("society", "urban planning"),
                    SkillOption("crafting", "construction")
                ),
                criticalSuccess = ActionEffect("Found village with bonus structure"),
                success = ActionEffect("Found village"),
                failure = ActionEffect("No effect"),
                criticalFailure = ActionEffect("No effect")
            ),
            PlayerAction(
                id = "build-structure",
                name = "Build Structure",
                category = "urban-planning",
                description = "Add markets, temples, barracks, and other structures",
                skills = listOf(
                    SkillOption("crafting", "construction"),
                    SkillOption("society", "urban development")
                ),
                criticalSuccess = ActionEffect("Build with reduced cost"),
                success = ActionEffect("Build structure"),
                failure = ActionEffect("No effect"),
                criticalFailure = ActionEffect("Waste resources")
            ),
            
            // Foreign Affairs
            PlayerAction(
                id = "establish-diplomatic-relations",
                name = "Establish Diplomatic Relations",
                category = "foreign-affairs",
                description = "Form alliances with other nations",
                skills = listOf(
                    SkillOption("diplomacy", "negotiation"),
                    SkillOption("society", "cultural exchange")
                ),
                criticalSuccess = ActionEffect("Strong alliance formed"),
                success = ActionEffect("Alliance formed"),
                failure = ActionEffect("No effect"),
                criticalFailure = ActionEffect("Diplomatic incident")
            ),
            PlayerAction(
                id = "hire-adventurers",
                name = "Hire Adventurers",
                category = "foreign-affairs",
                description = "Pay gold to resolve events (2 Gold cost)",
                skills = listOf(
                    SkillOption("diplomacy", "negotiate terms"),
                    SkillOption("society", "find adventurers")
                ),
                criticalSuccess = ActionEffect("Resolve event for 1 Gold"),
                success = ActionEffect("Resolve event for 2 Gold"),
                failure = ActionEffect("No effect"),
                criticalFailure = ActionEffect("Lose 2 Gold, no effect"),
                cost = mapOf("gold" to 2)
            ),
            
            // Economic Actions
            PlayerAction(
                id = "sell-surplus",
                name = "Sell Surplus",
                category = "economic-actions",
                description = "Trade resources for gold",
                skills = listOf(
                    SkillOption("diplomacy", "trade negotiations"),
                    SkillOption("society", "market knowledge")
                ),
                criticalSuccess = ActionEffect("Sell at premium prices"),
                success = ActionEffect("Sell at market prices"),
                failure = ActionEffect("No effect"),
                criticalFailure = ActionEffect("Bad deal, lose resources")
            ),
            PlayerAction(
                id = "collect-resources",
                name = "Collect Resources",
                category = "economic-actions",
                description = "Gather from hexes with or without worksites",
                skills = listOf(
                    SkillOption("survival", "resource gathering"),
                    SkillOption("nature", "identify resources")
                ),
                criticalSuccess = ActionEffect("Collect double resources"),
                success = ActionEffect("Collect resources"),
                failure = ActionEffect("No effect"),
                criticalFailure = ActionEffect("No effect")
            )
        )
    }
    
    fun getActionsByCategory(category: String): List<PlayerAction> {
        return getAllActions().filter { it.category == category }
    }
}

File: src-ts/models/PlayerActions.ts
// Player Actions data model for PF2e Kingdom Lite
// Auto-converted and fixed from PlayerActions.kt

/**
 * Represents a skill option for an action
 */
export interface SkillOption {
  skill: string;
  description: string;
}

/**
 * Represents the effect of an action outcome
 */
export interface ActionEffect {
  description: string;
  modifiers?: Map<string, number | string>;
}

/**
 * Represents a player action that can be taken during the kingdom turn
 */
export interface PlayerAction {
  id: string;
  name: string;
  category: string;
  description: string;
  skills: SkillOption[];
  criticalSuccess: ActionEffect;
  success: ActionEffect;
  failure: ActionEffect;
  criticalFailure: ActionEffect;
  proficiencyScaling?: Map<string, number> | null;
  special?: string | null;
  cost?: Map<string, number> | null; // For actions that have resource costs
}

/**
 * Player Actions data management
 */
export const PlayerActionsData = {
  // Category display names
  categoryNames: new Map([
    ['uphold-stability', 'Uphold Stability'],
    ['military-operations', 'Military Operations'],
    ['expand-borders', 'Expand the Borders'],
    ['urban-planning', 'Urban Planning'],
    ['foreign-affairs', 'Foreign Affairs'],
    ['economic-actions', 'Economic Actions']
  ]),
  
  // Category descriptions
  categoryDescriptions: new Map([
    ['uphold-stability', "Maintain the kingdom's cohesion by resolving crises and quelling unrest."],
    ['military-operations', 'War must be waged with steel and strategy.'],
    ['expand-borders', 'Seize new territory to grow your influence and resources.'],
    ['urban-planning', 'Your people need places to live, work, trade, and worship.'],
    ['foreign-affairs', 'No kingdom stands alone.'],
    ['economic-actions', 'Manage trade and personal wealth.']
  ]),
  
  /**
   * Get all available player actions
   */
  getAllActions(): PlayerAction[] {
    // This will be replaced with actual JSON loading
    return [
      // Expand Borders actions
      {
        id: 'claim-hexes',
        name: 'Claim Hexes',
        category: 'expand-borders',
        description: "Assert sovereignty over new territories, expanding your kingdom's borders into unclaimed lands",
        skills: [
          { skill: 'nature', description: 'harmonize with the land' },
          { skill: 'survival', description: 'establish frontier camps' },
          { skill: 'intimidation', description: 'force submission' },
          { skill: 'occultism', description: 'mystical claiming rituals' },
          { skill: 'religion', description: 'divine mandate' }
        ],
        criticalSuccess: { 
          description: 'Claim all targeted hexes +1 extra hex', 
          modifiers: new Map([['hexesClaimed', 'proficiency+1']]) 
        },
        success: { 
          description: 'Claim targeted hexes (based on proficiency)', 
          modifiers: new Map([['hexesClaimed', 'proficiency']]) 
        },
        failure: { description: 'No effect' },
        criticalFailure: { description: 'No effect' },
        proficiencyScaling: new Map([['trained', 1], ['expert', 1], ['master', 2], ['legendary', 3]]),
        special: '+2 circumstance bonus when claiming hexes adjacent to 3+ controlled hexes'
      },
      {
        id: 'build-roads',
        name: 'Build Roads',
        category: 'expand-borders',
        description: 'Construct pathways between settlements to improve trade, travel, and military movement',
        skills: [
          { skill: 'crafting', description: 'engineering expertise' },
          { skill: 'survival', description: 'pathfinding routes' },
          { skill: 'athletics', description: 'manual labor' },
          { skill: 'nature', description: 'work with terrain' }
        ],
        criticalSuccess: { 
          description: 'Build roads +1 hex', 
          modifiers: new Map([['roadsBuilt', 2]]) 
        },
        success: { 
          description: 'Build roads', 
          modifiers: new Map([['roadsBuilt', 1]]) 
        },
        failure: { description: 'No effect' },
        criticalFailure: { description: 'No effect' }
      },
      {
        id: 'send-scouts',
        name: 'Send Scouts',
        category: 'expand-borders',
        description: 'Learn about unexplored hexes',
        skills: [
          { skill: 'survival', description: 'wilderness navigation' },
          { skill: 'nature', description: 'track and explore' }
        ],
        criticalSuccess: { description: 'Scout 2 hexes' },
        success: { description: 'Scout 1 hex' },
        failure: { description: 'No effect' },
        criticalFailure: { description: 'No effect' }
      },
      {
        id: 'fortify-hex',
        name: 'Fortify Hex',
        category: 'expand-borders',
        description: 'Strengthen defensive positions',
        skills: [
          { skill: 'crafting', description: 'build fortifications' },
          { skill: 'warfare', description: 'strategic placement' }
        ],
        criticalSuccess: { description: 'Fortify hex with bonus' },
        success: { description: 'Fortify hex' },
        failure: { description: 'No effect' },
        criticalFailure: { description: 'No effect' }
      },
      {
        id: 'create-worksite',
        name: 'Create Worksite',
        category: 'expand-borders',
        description: 'Establish farms, mines, quarries, or lumber camps',
        skills: [
          { skill: 'crafting', description: 'build infrastructure' },
          { skill: 'survival', description: 'identify resources' }
        ],
        criticalSuccess: { description: 'Create worksite with bonus' },
        success: { description: 'Create worksite' },
        failure: { description: 'No effect' },
        criticalFailure: { description: 'No effect' }
      },
      // Uphold Stability actions
      {
        id: 'coordinated-effort',
        name: 'Coordinated Effort',
        category: 'uphold-stability',
        description: 'Two PCs work together on a single action with a bonus',
        skills: [
          { skill: 'diplomacy', description: 'coordinate efforts' },
          { skill: 'society', description: 'organize teamwork' }
        ],
        criticalSuccess: { description: 'Action succeeds with major bonus' },
        success: { description: 'Action succeeds with bonus' },
        failure: { description: 'No effect' },
        criticalFailure: { description: 'No effect' }
      },
      {
        id: 'arrest-dissidents',
        name: 'Arrest Dissidents',
        category: 'uphold-stability',
        description: 'Convert unrest into imprisoned unrest',
        skills: [
          { skill: 'intimidation', description: 'suppress dissent' },
          { skill: 'society', description: 'identify troublemakers' }
        ],
        criticalSuccess: { description: 'Convert 2 unrest to imprisoned' },
        success: { description: 'Convert 1 unrest to imprisoned' },
        failure: { description: 'No effect' },
        criticalFailure: { description: 'Gain 1 unrest' }
      },
      {
        id: 'deal-with-unrest',
        name: 'Deal with Unrest',
        category: 'uphold-stability',
        description: 'Directly reduce unrest by 1-3 based on success',
        skills: [
          { skill: 'diplomacy', description: 'calm the populace' },
          { skill: 'performance', description: 'inspire the people' },
          { skill: 'intimidation', description: 'quell dissent' }
        ],
        criticalSuccess: { description: 'Reduce unrest by 3' },
        success: { description: 'Reduce unrest by 1' },
        failure: { description: 'No effect' },
        criticalFailure: { description: 'Gain 1 unrest' }
      },
      // Military Operations
      {
        id: 'recruit-unit',
        name: 'Recruit a Unit',
        category: 'military-operations',
        description: 'Raise new troops for your armies',
        skills: [
          { skill: 'warfare', description: 'military recruitment' },
          { skill: 'intimidation', description: 'conscription' }
        ],
        criticalSuccess: { description: 'Recruit unit with bonus' },
        success: { description: 'Recruit unit' },
        failure: { description: 'No effect' },
        criticalFailure: { description: 'No effect' }
      },
      {
        id: 'deploy-army',
        name: 'Deploy Army',
        category: 'military-operations',
        description: 'Move troops to strategic positions',
        skills: [
          { skill: 'warfare', description: 'tactical deployment' },
          { skill: 'survival', description: 'logistics' }
        ],
        criticalSuccess: { description: 'Deploy with extra movement' },
        success: { description: 'Deploy army' },
        failure: { description: 'No effect' },
        criticalFailure: { description: 'No effect' }
      },
      // Urban Planning
      {
        id: 'establish-settlement',
        name: 'Establish a Settlement',
        category: 'urban-planning',
        description: 'Found a new village',
        skills: [
          { skill: 'society', description: 'urban planning' },
          { skill: 'crafting', description: 'construction' }
        ],
        criticalSuccess: { description: 'Found village with bonus structure' },
        success: { description: 'Found village' },
        failure: { description: 'No effect' },
        criticalFailure: { description: 'No effect' }
      },
      {
        id: 'build-structure',
        name: 'Build Structure',
        category: 'urban-planning',
        description: 'Add markets, temples, barracks, and other structures',
        skills: [
          { skill: 'crafting', description: 'construction' },
          { skill: 'society', description: 'urban development' }
        ],
        criticalSuccess: { description: 'Build with reduced cost' },
        success: { description: 'Build structure' },
        failure: { description: 'No effect' },
        criticalFailure: { description: 'Waste resources' }
      },
      // Foreign Affairs
      {
        id: 'establish-diplomatic-relations',
        name: 'Establish Diplomatic Relations',
        category: 'foreign-affairs',
        description: 'Form alliances with other nations',
        skills: [
          { skill: 'diplomacy', description: 'negotiation' },
          { skill: 'society', description: 'cultural exchange' }
        ],
        criticalSuccess: { description: 'Strong alliance formed' },
        success: { description: 'Alliance formed' },
        failure: { description: 'No effect' },
        criticalFailure: { description: 'Diplomatic incident' }
      },
      {
        id: 'hire-adventurers',
        name: 'Hire Adventurers',
        category: 'foreign-affairs',
        description: 'Pay gold to resolve events (2 Gold cost)',
        skills: [
          { skill: 'diplomacy', description: 'negotiate terms' },
          { skill: 'society', description: 'find adventurers' }
        ],
        criticalSuccess: { description: 'Resolve event for 1 Gold' },
        success: { description: 'Resolve event for 2 Gold' },
        failure: { description: 'No effect' },
        criticalFailure: { description: 'Lose 2 Gold, no effect' },
        cost: new Map([['gold', 2]])
      },
      // Economic Actions
      {
        id: 'sell-surplus',
        name: 'Sell Surplus',
        category: 'economic-actions',
        description: 'Trade resources for gold',
        skills: [
          { skill: 'diplomacy', description: 'trade negotiations' },
          { skill: 'society', description: 'market knowledge' }
        ],
        criticalSuccess: { description: 'Sell at premium prices' },
        success: { description: 'Sell at market prices' },
        failure: { description: 'No effect' },
        criticalFailure: { description: 'Bad deal, lose resources' }
      },
      {
        id: 'collect-resources',
        name: 'Collect Resources',
        category: 'economic-actions',
        description: 'Gather from hexes with or without worksites',
        skills: [
          { skill: 'survival', description: 'resource gathering' },
          { skill: 'nature', description: 'identify resources' }
        ],
        criticalSuccess: { description: 'Collect double resources' },
        success: { description: 'Collect resources' },
        failure: { description: 'No effect' },
        criticalFailure: { description: 'No effect' }
      }
    ];
  },
  
  /**
   * Get actions by category
   */
  getActionsByCategory(category: string): PlayerAction[] {
    return this.getAllActions().filter(action => action.category === category);
  },
  
  /**
   * Get category display name
   */
  getCategoryName(category: string): string {
    return this.categoryNames.get(category) || category;
  },
  
  /**
   * Get category description
   */
  getCategoryDescription(category: string): string {
    return this.categoryDescriptions.get(category) || '';
  }
};
